GitUI/UserControls/ExListView.cs

```diff
@@ -72,14 +76,50 @@ internal class ExListView : NativeListView
 
         static ExListView()
         {
-            ListViewGroupIdProperty = typeof(ListViewGroup).GetProperty("ID", BindingFlags.GetProperty | BindingFlags.Instance | BindingFlags.NonPublic);
+            ListViewGroupIdProperty = typeof(ListViewGroup).GetProperty("ID",
+                BindingFlags.GetProperty | BindingFlags.Instance | BindingFlags.NonPublic);
+
+            ListViewGroupListProperty = typeof(ListViewGroupCollection).GetProperty("List",
+                BindingFlags.GetProperty | BindingFlags.Instance | BindingFlags.NonPublic);
+
+            ListViewDefaultGroupProperty = typeof(ListView).GetProperty("DefaultGroup",
+                BindingFlags.GetProperty | BindingFlags.Instance | BindingFlags.NonPublic);
         }
 
         public ExListView()
         {
             DoubleBuffered = true;
+            IsGroupStateSupported = EnvUtils.RunningOnWindows() && Environment.OSVersion.Version.Major >= 6;
+        }
+
+        /// <summary>
+        /// Call this method once after <see cref="ListView.Groups"/> collection was cleared (or created)
+        /// before making any calls to <see cref="ListViewGroupCollection.Insert"/>
+        ///
+        /// <see cref="ListViewGroupCollection.Add(System.Windows.Forms.ListViewGroup)"/> calls must
+        /// not be used after calling this method
+        /// </summary>
+        public void WorkaroundGroupInsertionBug()
+        {
+            // .NET ListView.Groups.Insert(...) implementation has a bug
+            // It does not count the technical "Default" group from ListView when passing inserted
+            // group index to native Win32 ListView.
+
+            // ListView.Groups.Add(...) implementation on the other hand does count the
+            // technical "Default" group correctly therefore it becomes broken after calling this method
+
+            var defaultGroup = ListViewDefaultGroupProperty.GetValue(this);
+            var list = (ArrayList)ListViewGroupListProperty.GetValue(Groups);
+            list.Insert(0, defaultGroup);
+            MinGroupInsertionIndex = 1;
         }
 
+        /// <summary>
+        /// Position in <see cref="ListView.Groups"/> collection after
+        /// the technical "Default" group (if present) or 0
+        /// </summary>
+        public int MinGroupInsertionIndex { get; private set; } = 0;
```

Comment:
This seems like an implementation detail, somewhat irrelevant and an overhead for consumers to know or care.
Could we perhaps restructure the code to offer API like this (pseudocode writing on a train):

```
public void InsertGroup(ListViewGroup group) { ... }
public void InsertGroups(IEnumerable<ListViewGroup> groups) { ... }
```

I'd like to turn `WorkaroundGroupInsertionBug` into `private`, if possible, too.

Reply:
I wanted to do it completely incapsulated from the very beginnig. And additionally, I wanted to hide Groups collection by a new member with the same name. It would delegate calls to actual Groups collection and handle the workaround.

What stopped me was the fact that it would break using `ExListVIew.Groups` collection in `.designer.cs` files, where some groups are added for the sake of visual assistance in design-time.

Do you think we can sacrifice these usages?

Reply:
Created `SetGroups` methods in `ExListView` which teporarily enables the workaround to do its job hiding workaround details from client code.